class Solution {
    public boolean checkValidString(String s) {
        int minOpen = 0;
        int maxOpen = 0;

        for (char ch : s.toCharArray()) {

            if (ch == '(') {
                minOpen++;
                maxOpen++;
            } 
            else if (ch == ')') {
                minOpen--;
                maxOpen--;
            } 
            else { // '*'
                minOpen--;
                maxOpen++;
            }

            // We cannot have fewer than 0 unmatched '('
            minOpen = Math.max(0, minOpen);

            // Even the maximum possible '(' count is negative
            if (maxOpen < 0) {
                return false;
            }
        }

        return minOpen == 0;
    }
}